testers.nixosTest: Deprecate

roberth opened the pull request:
## Description of changes

It's time to migrate.
`runNixOSTest` is a better alternative, that supports modules properly, as well as running on macOS.
At this point, the existence of `nixosTest` leads to uncertainty, and its users are missing out on the new features.

:rocket: 

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
testers.nixosTest: Deprecate

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixostest-vs-runtest/26419/6

AndersonTorres commented:
Why is this not working?

DanielSidhion commented on doc/build-helpers/testers.chapter.md:
```diff
@@ -229,6 +229,12 @@ pkgs.testers.runNixOSTest ({ lib, ... }: {
 
 ## `nixosTest` {#tester-nixosTest}
 
+This function is deprecated in favor of [`runNixOSTest`](#tester-runNixOSTest). The reason why it's bad is that when you pass it a function, it is invoked via `callPackage`. This is somewhat incompatible with the module system and running cross-platform:
+  - The choice of packages is ambiguous when the VM host platform is not Linux. `nixosTest` could in theory pick either:
+    - `hostPkgs` for packages to run on the VM host
+    - or `config.node.pkgs` for packages that run in one of the VMs
+  - The return value of the passed function should have been a module, but `nixosTest` does not actually allow that.
+
```
I think you should wrap it in a warning box to grab more attention.
Also moved each sentence to its own line.

```suggestion
:::{.warning}
This function is deprecated in favor of [`runNixOSTest`](#tester-runNixOSTest).
The reason why it's bad is that when you pass it a function, it is invoked via `callPackage`.
This is somewhat incompatible with the module system and running cross-platform:
  - The choice of packages is ambiguous when the VM host platform is not Linux.
    `nixosTest` could in theory pick either:
    - `hostPkgs` for packages to run on the VM host
    - or `config.node.pkgs` for packages that run in one of the VMs
  - The return value of the passed function should have been a module, but `nixosTest` does not actually allow that.
:::
```

roberth committed:
doc/testers: Edit deprecation warning for nixosTest

Co-authored-by: Daniel Sidhion <DanielSidhion@users.noreply.github.com>

roberth commented:
> Why is this not working?

`nixosTest` is not completely broken, but has issues, as described.

Maybe I don't understand the question, but I think I've explained that in the diff, and I've added extra context in the PR description.

AndersonTorres commented:
I was talking about the ofborg red status.

DanielSidhion reviewed (approved):
LGTM on the docs changes. Idk what's the reasoning behind the ofborg red status though.

infinisil commented on pkgs/build-support/testers/default.nix:
```diff
@@ -118,6 +118,7 @@
        * make sense for this evaluation of Nixpkgs.
        */
       nixosTesting =
+        lib.warn "testers.nixosTest is deprecated. Use testers.runNixOSTest instead."
```
The ofborg failure could be fixed by moving this warning into the function body under `test:`. Non ideal that this is necessary though..

infinisil replied:
Never mind, looks like that doesn't work either..

Thinking about this again, I believe we need to not expose the attribute at all if `config.allowAliases = false` (which then also prevents uses in Nixpkgs).

roberth committed:
testers.nixosTest: Move warning and add link

This way ofborg doesn't see it and fail.
Unfortunately, it will cause more warnings than it otherwise
would, if multiple tests are evaluated in one go.

roberth committed:
testers.nixosTests: Delete unless allowAliases

roberth committed:
treewide: nixosTest -> runNixOSTest, or remove

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixostest-with-flake-configurations/11542/8
